fix for issue #10 - Adding unit tests

codemedic opened the pull request:
Also:
- GPG now uses AES256 and is backward compatible
- GPG is quieter (was a lil noisy bugger)
- reduced info logging (made them debugs)
- some minor refactoring

codemedic committed:
fix for issue #10 - Adding unit tests

Also:
 - GPG now uses AES256 and is backward compatible
 - GPG is quieter (was a lil noisy bugger)
 - reduced info logging (made them debugs)
 - some minor refactoring

codemedic committed:
issue #10 - updating docs

abulford commented on s3sync.sh:
```diff
@@ -3,12 +3,16 @@
 
 s3sync() {
+    is_enabled() {
+        [ -n "${AWS_ACCESS_KEY}${AWS_SECRET_KEY}" ]
```
This will return true, indicating that S3 sync is enabled, even if not both access key and secret key are set, which would be invalid.

abulford commented on entrypoint.sh:
```diff
@@ -81,7 +67,7 @@ restore() {
     local gpg_file="$ENCRYPTED_PATH/$1"
     if file_not_exists "$gpg_file"; then
-        if is_aws_s3_enabled; then
+        if ! s3sync is_enabled; then
```
Isn't this backwards?  I would think if it's not enabled then you wouldn't attempt to sync.  Could do with an `else` for the scenario where it's not enabled.

abulford commented on entrypoint.sh:
```diff
@@ -107,27 +93,17 @@ restore() {
     local batch_mode=$([[ "$1" = "--batch-mode" ]] && echo true || echo false)
     local files_in_s3=();
-    if is_aws_s3_enabled; then
+    if s3sync is_enabled; then
```
Maybe an `else` here as well?  Or maybe instead of all the `else` statements the `is_enabled` function could print the message to say it's not enabled.

codemedic replied:
This and the other .. I have rejigged a bit so that the script will fail before it does anything.
At that point, it will also print out an `info` line if the functionality is not configured .. i.e all variables are empty.

abulford commented on functions.sh:
```diff
@@ -85,4 +85,24 @@ load_settings_file() {
     fi
 }
 
+# in_array - check if an element can be found in an array
```
It doesn't seem right to define a function like this which does such a generic thing - shouldn't it come from a library or something?  Or is this indicating that bash isn't the right choice of language for this application?

codemedic replied:
:astonished: Donno what to say andy!

codemedic committed:
Fix issue #12 - Verifies before sync and decrypt

Fix comments from PR #11
 - Stricter AWS S3 env variable checks
 - Added test for #12
 - Minor typo fixes (from #11)

codemedic commented:
@abulford 
I have updated the PR; please do have a look.

abulford commented on run-test.sh:
```diff
@@ -0,0 +1,43 @@
+        local filepath="${SOURCE_PATH}/data-$(printf %02d $i).txt"
+        local filename=$(basename "$filepath")
+
+        if is_number "$SOURCE_REMOVE_PLAIN" && [ "$SOURCE_REMOVE_PLAIN" -eq 1 ]; then
```
Any reason not to do `if [ "${SOURCE_REMOVE_PLAIN} == "1" ]` ?

abulford commented on run-test.sh:
```diff
@@ -0,0 +1,43 @@
+            bail_file_exists "$filepath";
+        fi
+
+        /app/entrypoint.sh restore "${filename}.gpg"
```
Where does the `.gpg` file come from?  I'm sure it must work, but without digging through the code I can't figure out how the test should pass when `entrypoint.sh` is being asked to restore a file which doesn't appear to have been prepared within the script.

abulford commented on run-test.sh:
```diff
@@ -5,8 +5,9 @@
 source /app/functions.sh
+source /app/s3sync.sh
 
-set -e;
```
If `set -e` is removed, will the `verify_test_file` below cause the script to exit if it fails?

codemedic replied:
unintended ... will fix that!

abulford commented on run-test.sh:
```diff
@@ -26,17 +27,31 @@ verify_test_file() {
 
+        # if s3sync is enabled, then remove local encrypted files so that they are pulled down from S3
+        if s3sync is_enabled; then
+            rm "${ENCRYPTED_PATH}/"*.txt.gpg
```
Should this go outside the loop?

codemedic replied:
I think it should ... will fix that

abulford commented on s3sync.sh:
```diff
@@ -10,7 +10,15 @@ check_command awk
+        # if any AWS S3 tingz are set, then they all should be set
+        if [ -n "${AWS_ACCESS_KEY}${AWS_SECRET_KEY}${AWS_S3_BUCKET}" ]; then
+            if [ -z "${AWS_ACCESS_KEY}" -o -z "${AWS_SECRET_KEY}" -o -z "${AWS_S3_BUCKET}" ]; then
+                bail "s3sync: Not all AWS S3 config values are specifies"
```
Typo: 'specifies' -> 'specified'

codemedic committed:
Fixing comments raised in #11
